src/Test.UnitTests.Sarif/EnumeratedArtifactTests.cs

```diff
@@ -16,7 +20,223 @@ namespace Test.UnitTests.Sarif
+            // enumerated artifact is operating strictly from the stream.
+            tempFile.Dispose();
+            File.Exists(filePath).Should().BeFalse();
+
+            ValidateTextArtifact(artifact, fileSize);
+        }
+
+        [Fact]
+        public void EnumeratedArtifact_TextFile_NonSeekableStream()
+        {
+            string guid = $"{Guid.NewGuid()}";
+            ZipArchive archive = CreateZipArchive("MyTextualFile.txt", Encoding.UTF8.GetBytes(guid));
+
+            foreach (ZipArchiveEntry entry in archive.Entries)
+            {
+                var artifact = new EnumeratedArtifact(FileSystem.Instance)
+                {
+                    Uri = new Uri(entry.FullName, UriKind.RelativeOrAbsolute),
+                    SupportNonSeekableStreams = true,
+                    Stream = entry.Open(),
+                };
+
+                ValidateTextArtifact(artifact, guid.Length);
+            }
+        }
+
+        [Fact]
+        public void EnumeratedArtifact_BinaryFile_NonSeekableStream()
+        {
+            string filePath = this.GetType().Assembly.Location;
+            using FileStream reader = File.OpenRead(filePath);
+
+            int headerSize = 1024;
+            byte[] data = new byte[headerSize];
+            reader.Read(data, 0, data.Length);
+
+            ZipArchive archive = CreateZipArchive("MyBinaryFile.dll", data);
+
+            foreach (ZipArchiveEntry entry in archive.Entries)
+            {
+                var artifact = new EnumeratedArtifact(FileSystem.Instance)
+                {
+                    Uri = new Uri(entry.FullName, UriKind.RelativeOrAbsolute),
+                    SupportNonSeekableStreams = true,
+                    Stream = entry.Open(),
+                };
+
+                ValidateBinaryArtifact(artifact, headerSize);
+            }
+        }
+
+        [Fact]
+        public void EnumeratedArtifact_TextFile_SizeInBytesWithRelativeUriReturnsNull()
+        {
+            var artifact = new EnumeratedArtifact(new FileSystem())
+            {
+                Uri = new Uri("/test.txt", UriKind.Relative)
+            };
+
+            Assert.Throws<InvalidOperationException>(() => artifact.SizeInBytes);
+        }
+
+        [Fact]
+        public void EnumeratedArtifact_TextFile_NonSeekableStreamRaiseExceptions()
+        {
+            string guid = $"{Guid.NewGuid()}";
+            ZipArchive archive = CreateZipArchive("MyTextualFile.txt", Encoding.UTF8.GetBytes(guid));
+
+            foreach (ZipArchiveEntry entry in archive.Entries)
+            {
+                var artifact = new EnumeratedArtifact(FileSystem.Instance)
+                {
+                    Stream = entry.Open(),
+                };
+
+                Assert.Throws<InvalidOperationException>(() => ValidateTextArtifact(artifact, guid.Length));
+            }
+        }
+
+        private void ValidateBinaryArtifact(EnumeratedArtifact artifact, int sizeInBytes)
```

Comment:
Inferring standards from practice: are we okay with sequences of assertions wherever the [Fact] is validating a single input/case, as opposed to a multiplicity of them?

I'm hypothesizing that if this method were ValidateBinaryArtifact**s** then it would be using a StringBuilder.

Reply:
Yes! For a true, more atomic unit test, it's fine to break early on the first condition that fails. There's two interesting things to consider, one is what failure information is more useful to the debugger. The second is to ensure that the literal output of the test is illustrative.

Some teams have stringent standards around emitting textual messages with all failures. Seeing 'Expected '17' but observed '13'' isn't always helpful.